File: apps/server/dekobloko_server/js5.py
from __future__ import annotations

from pathlib import Path

import socket
import struct
import zlib
from dataclasses import dataclass

from .cache import CacheStore
from .io import read_exact, read_u8


# --- substitutes for cache groups that no longer exist anywhere -------------
#
# Archive 6 group 1 is "benefits" (files: borders, logo, price, screenshots) --
# the FunOrb subscription upsell panel. The original service is dead and the
# data is in no surviving cache, so the client blocks forever on "Loading extra
# data" waiting for it. An empty container does NOT satisfy it: the client
# treats a zero-length group as a failed fetch and re-requests indefinitely.
#
# So serve a structurally valid replacement instead. See
# docs/js5-sprite-format.md; the generator is synthetic/make_happy.py, a
# sibling of the dekobloko_server package (apps/server/synthetic/).
_SUBSTITUTE_DIR = Path(__file__).resolve().parents[1] / "synthetic"
_SUBSTITUTES = {(6, 1): "archive6_group1.bin"}
_substitute_cache: dict[tuple[int, int], bytes | None] = {}


def _load_substitute(archive_id: int, group_id: int) -> bytes | None:
    key = (archive_id, group_id)
    if key in _substitute_cache:
        return _substitute_cache[key]
    name = _SUBSTITUTES.get(key)
    data = None
    if name is not None:
        path = _SUBSTITUTE_DIR / name
        try:
            data = path.read_bytes()
        except OSError as exc:
            print(f"[js5] substitute {archive_id}/{group_id} unreadable at {path}: {exc}")
    _substitute_cache[key] = data
    return data


# --- CRC reconciliation for substituted groups -----------------------------
#
# The client validates every group it downloads against the CRC recorded for it
# in that archive's group table, and rejects a mismatch with "CRC mismatch -
# unable to get a valid download" plus an endless Retry loop. A synthesised
# group can never match the CRC of the data it replaces, and the CRC cannot be
# worked backwards into the original bytes -- 32 bits against ~10^5 bytes.
#
# We serve the group table too, so the fix runs the other way: rewrite the
# recorded CRC to match what we actually send. That is a two-level chain,
# because the master index (255/255) records a CRC over each group table:
#
#     substitute bytes      -> CRC lands in archive N's group table (255/N)
#     patched 255/N         -> CRC lands in the master index (255/255)
#
# Both levels must go through _patched_read(), or the client accepts the group
# and then rejects the table that vouches for it.
#
# Verified against this cache: the CRC covers the whole container, header
# included, and there is no version trailer on the 255/N entries.

def _decompress(raw: bytes) -> bytes:
    compression = raw[0]
    length = int.from_bytes(raw[1:5], "big")
    if compression == 0:
        return raw[5:5 + length]
    body = raw[9:9 + length]
    if compression == 1:
        import bz2
        return bz2.decompress(b"BZh1" + body)
    import gzip
    import io as _io
    return gzip.GzipFile(fileobj=_io.BytesIO(body)).read()


def _as_container(payload: bytes) -> bytes:
    """Wrap payload as an uncompressed js5 container.

    Re-emitting uncompressed rather than recompressing keeps the bytes
    deterministic, which matters because the CRC we then advertise must match
    exactly what goes on the wire.
    """
    return bytes([0]) + len(payload).to_bytes(4, "big") + payload


def _group_table_crc_offset(table: bytes, group_id: int) -> int | None:
    """Byte offset of group_id's CRC field inside a decompressed group table."""
    p = 0
    protocol = table[p]; p += 1
    if protocol >= 6:
        p += 4                                   # version
    flags = table[p]; p += 1
    count = int.from_bytes(table[p:p + 2], "big"); p += 2
    ids = []
    acc = 0
    for _ in range(count):
        acc += int.from_bytes(table[p:p + 2], "big"); p += 2
        ids.append(acc)
    if flags & 1:
        p += 4 * count                           # name hashes
    if group_id not in ids:
        return None
    return p + 4 * ids.index(group_id)


@dataclass(frozen=True)
class Js5Request:
    priority: bool
    archive_id: int
    group_id: int


class Js5Session:
    def __init__(self, sock: socket.socket, cache: CacheStore, peer: str) -> None:
        self.sock = sock
        self.cache = cache
        self.peer = peer
        self.xor_key = 0

    def run_after_handshake(self, revision: int) -> None:
        print(f"[js5] {self.peer} accepted revision={revision}")
        self.sock.sendall(b"\x00")
        while True:
            packet = read_exact(self.sock, 6)
            opcode = packet[0]

            if opcode in (0, 1):
                request = Js5Request(
                    priority=opcode == 1,
                    archive_id=packet[1],
                    group_id=int.from_bytes(packet[2:6], "big"),
                )
                self._handle_file_request(request)
                continue

            if opcode == 2:
                print(f"[js5] {self.peer} priority control {packet.hex(' ')}")
                continue

            if opcode == 3:
                print(f"[js5] {self.peer} normal control {packet.hex(' ')}")
                continue

            if opcode == 4:
                self.xor_key = packet[1]
                print(f"[js5] {self.peer} xor-key={self.xor_key}")
                continue

            if opcode == 6:
                print(f"[js5] {self.peer} setup {packet.hex(' ')}")
                continue

            if opcode == 7:
                print(f"[js5] {self.peer} disconnect control")
                return

            print(f"[js5] {self.peer} unknown opcode={opcode} packet={packet.hex(' ')}")

    def _patched_read(self, archive_id: int, group_id: int) -> bytes | None:
        """cache.read() with substituted groups and reconciled CRCs applied.

        Every read path must use this. Reading the cache directly anywhere
        re-introduces the mismatch this exists to remove.
        """
        raw = self.cache.read(archive_id, group_id)

        # Level 0: the substituted group itself.
        if raw is None:
            substitute = _load_substitute(archive_id, group_id)
            if substitute is not None:
                print(
                    f"[js5] {self.peer} SUBSTITUTE archive={archive_id} "
                    f"group={group_id} bytes={len(substitute)} "
                    f"crc=0x{zlib.crc32(substitute) & 0xFFFFFFFF:08x}"
                )
                return substitute
            return None

        # Levels 1 and 2 are DISABLED. Rewriting the recorded CRC does not work:
        # the master index (255/255) is *signed*, and it covers the group-table
        # CRCs. Patching a group table forces a matching edit to the master
        # index, which invalidates the signature -- the client verifies it,
        # throws RuntimeException and dies with error_game_crash before it ever
        # reaches the login screen.
        #
        # The workable direction is the reverse: forge the substitute so its CRC
        # equals the value already recorded (0xaacfba29 for archive 6 group 1).
        # CRC32 is affine over GF(2), so 32 probe CRCs are enough to solve for
        # four bytes that land on any target. Put them somewhere inert -- the
        # palette, not pixel indices, which could go out of bounds and crash the
        # sprite reader. Then nothing signed has to change.
        return raw

        if archive_id != 255:
            return raw

        # Level 1: a group table vouching for a substituted group.
        if group_id != 255:
            patched = raw
            for (arch, grp), _name in _SUBSTITUTES.items():
                if arch != group_id:
                    continue
                substitute = _load_substitute(arch, grp)
                if substitute is None:
                    continue
                table = bytearray(_decompress(patched))
                off = _group_table_crc_offset(bytes(table), grp)
                if off is None:
                    print(f"[js5] {self.peer} group {grp} absent from table {arch}; CRC not patched")
                    continue
                want = zlib.crc32(substitute) & 0xFFFFFFFF
                had = int.from_bytes(table[off:off + 4], "big")
                table[off:off + 4] = want.to_bytes(4, "big")
                patched = _as_container(bytes(table))
                print(
                    f"[js5] {self.peer} patched table {arch} group {grp} "
                    f"crc 0x{had:08x} -> 0x{want:08x}"
                )
            return patched

        # Level 2: the master index vouching for those group tables.
        touched = {arch for arch, _ in _SUBSTITUTES}
        if not touched:
            return raw
        body = bytearray(_decompress(raw))
        count = body[0]
        for arch in sorted(touched):
            if arch >= count:
                continue
            table = self._patched_read(255, arch)
            if table is None:
                continue
            off = 1 + arch * 72                  # crc(4) + version(4) + whirlpool(64)
            want = zlib.crc32(table) & 0xFFFFFFFF
            had = int.from_bytes(body[off:off + 4], "big")
            if had == want:
                continue
            body[off:off + 4] = want.to_bytes(4, "big")
            print(
                f"[js5] {self.peer} patched master index archive {arch} "
                f"crc 0x{had:08x} -> 0x{want:08x}"
            )
        return _as_container(bytes(body))

    def _build_master_index(self) -> bytes:
        """Generate the js5 master index (archive 255, group 255) from the per-archive
        indexes (255, N). The master index is never stored on disk; a real js5 server
        computes it live. Format: container([count:u8] then per archive
        [crc:u32][version:u32][whirlpool:64b])."""
        idx_path = self.cache.cache_dir / "main_file_cache.idx255"
        count = idx_path.stat().st_size // 6 if idx_path.is_file() else 0
        body = bytearray([count & 0xFF])
        for n in range(count):
            raw = self._patched_read(255, n)
            if raw is None:
                crc = 0
            else:
                crc = zlib.crc32(raw) & 0xFFFFFFFF
            body.extend(struct.pack(">II", crc, 0))  # crc, version=0
            body.extend(b"\x00" * 64)                # whirlpool (zero-filled)
        return bytes([0]) + struct.pack(">I", len(body)) + bytes(body)  # compression 0

    def _handle_file_request(self, request: Js5Request) -> None:
        if request.archive_id == 255 and request.group_id == 255:
            raw = self._patched_read(255, 255)
            if raw is None:
                raw = self._build_master_index()
                print(f"[js5] {self.peer} generated unsigned fallback master index ({len(raw)}b)")
            else:
                print(f"[js5] {self.peer} loaded signed master index ({len(raw)}b)")
        else:
            raw = self._patched_read(request.archive_id, request.group_id)
        if raw is None:
            print(
                f"[js5] {self.peer} MISS archive={request.archive_id} "
                f"group={request.group_id} priority={request.priority}"
            )
            # Answer the miss instead of staying silent.
            #
            # A bare return leaves the client waiting for a response that never
            # arrives; it holds the JS5 connection open, eventually drops it, and
            # reopens ~30s later, forever. Observed at game-over: the client asks
            # for archive=6 group=1 and archive=2 group=3, neither of which is in
            # any local cache, so it never returns to the main menu and looks
            # hung. Client instrumentation showed byte-identical request bursts
            # exactly 30s apart with no server reply in between.
            #
            # Replying with a well-formed empty container (compression 0, length
            # 0) at least lets the client's reader complete and decide for itself
            # what to do with an empty group, rather than blocking on silence.
            # This does NOT conjure up the missing data — if the client genuinely
            # needs those bytes it will fail some other way, but it should fail
            # visibly instead of hanging.
            flags = 0
            if not request.priority:
                flags |= 0x80
            empty = bytearray()
            empty.append(request.archive_id & 0xFF)
            empty.extend(request.group_id.to_bytes(4, "big"))
            empty.append(flags)
            empty.extend((0).to_bytes(4, "big"))
            self._send_framed(bytes(empty))
            print(
                f"[js5] {self.peer} sent EMPTY container for miss "
                f"archive={request.archive_id} group={request.group_id}"
            )
            return

        if request.archive_id != 255:
            compression = raw[0]
            compressed_length = int.from_bytes(raw[1:5], "big")
            container_length = (5 if compression == 0 else 9) + compressed_length
            if len(raw) < container_length:
                print(
                    f"[js5] {self.peer} truncated cache entry archive={request.archive_id} "
                    f"group={request.group_id} bytes={len(raw)} expected={container_length}"
                )
                return
            if len(raw) > container_length:
                trailer = raw[container_length:]
                print(
                    f"[js5] {self.peer} stripped disk trailer={trailer.hex()} "
                    f"archive={request.archive_id} group={request.group_id}"
                )
                raw = raw[:container_length]

        if len(raw) < 5:
            print(f"[js5] {self.peer} invalid short cache entry: {request}")
            return

        compression = raw[0]
        compressed_length = int.from_bytes(raw[1:5], "big")
        flags = compression & 0x7F
        if not request.priority:
            flags |= 0x80

        packet = bytearray()
        packet.append(request.archive_id & 0xFF)
        packet.extend(request.group_id.to_bytes(4, "big"))
        packet.append(flags)
        packet.extend(compressed_length.to_bytes(4, "big"))
        packet.extend(raw[5:])

        self._send_framed(bytes(packet))
        print(
            f"[js5] {self.peer} sent archive={request.archive_id} "
            f"group={request.group_id} bytes={len(raw)} priority={request.priority}"
        )

    def _xor(self, data: bytes) -> bytes:
        """Apply the client-selected JS5 XOR key to outgoing bytes.

        The client may send opcode 4 with a key byte, after which it expects
        every subsequent byte of the JS5 stream — headers, framing 0xFF markers
        and payload alike — to be XORed with that key. Storing the key but
        sending plaintext makes the client decode garbage and drop the
        connection with "IO error - unable to communicate reliably with the data
        server". Key 0 is the identity, i.e. XOR disabled.
        """
        if not self.xor_key:
            return data
        k = self.xor_key & 0xFF
        return bytes(b ^ k for b in data)

    def _send_framed(self, packet: bytes) -> None:
        if len(packet) <= 512:
            self.sock.sendall(self._xor(packet))
            return

        offset = 512
        self.sock.sendall(self._xor(packet[:offset]))
        while offset < len(packet):
            self.sock.sendall(self._xor(b"\xff"))
            end = min(offset + 511, len(packet))
            self.sock.sendall(self._xor(packet[offset:end]))
            offset = end

File: apps/server-js/src/js5.js
"use strict";
// Port of dekobloko_server/js5.py (JS5 handshake + archive/group request
// serving, including the CRC-patch trickery documented inline below).
//
// Python methods wrote to a blocking socket; here the network-facing methods
// are async and write via sock.write() callbacks. The pure byte-shaping logic
// (build_file_packet / frame_packet / xor_bytes) is factored into exported
// functions so golden-vector tests can exercise it without sockets -- the
// session methods are thin orchestration over the same code.

const fs = require("fs");
const path = require("path");

const { bzip2_decompress_bzh1, crc32 } = require("./bzip2.js");

// --- substitutes for cache groups that no longer exist anywhere -------------
//
// Archive 6 group 1 is "benefits" (files: borders, logo, price, screenshots) --
// the FunOrb subscription upsell panel. The original service is dead and the
// data is in no surviving cache, so the client blocks forever on "Loading extra
// data" waiting for it. An empty container does NOT satisfy it: the client
// treats a zero-length group as a failed fetch and re-requests indefinitely.
//
// So serve a structurally valid replacement instead. The generator lives in
// apps/server/synthetic/ (make_happy.py); Python resolved that directory as
// Path(__file__).parents[1] / "synthetic", which from src/js5.js maps to
// ../../server/synthetic.
const SUBSTITUTE_DIR =
  process.env.DEKOBLOKO_SYNTHETIC_DIR !== undefined
    ? process.env.DEKOBLOKO_SYNTHETIC_DIR
    : path.join(__dirname, "..", "..", "server", "synthetic");
const _SUBSTITUTES = new Map([["6,1", "archive6_group1.bin"]]);
const _substitute_cache = new Map(); // "archive,group" -> Buffer | null

function _key(archive_id, group_id) {
  return `${archive_id},${group_id}`;
}

function _load_substitute(archive_id, group_id) {
  const key = _key(archive_id, group_id);
  if (_substitute_cache.has(key)) {
    return _substitute_cache.get(key);
  }
  const name = _SUBSTITUTES.get(key);
  let data = null;
  if (name !== undefined) {
    const p = path.join(SUBSTITUTE_DIR, name);
    try {
      data = fs.readFileSync(p);
    } catch (err) {
      console.log(`[js5] substitute ${archive_id}/${group_id} unreadable at ${p}: ${err.message}`);
    }
  }
  _substitute_cache.set(key, data);
  return data;
}

// --- CRC reconciliation for substituted groups -----------------------------
//
// (Full rationale in apps/server/dekobloko_server/js5.py.) The client checks
// every downloaded group against the CRC in its archive's group table, and the
// master index signs those tables -- so a naive patch chain invalidates the
// signature and kills the client with error_game_crash. Levels 1 and 2 below
// are therefore DISABLED in Python (unreachable after an unconditional
// return); the working approach forges the substitute so its CRC already
// equals the recorded value (0xaacfba29 for archive 6 group 1). The helpers
// and disabled branches are ported anyway so diffs against Python stay
// mechanical.

function _decompress(raw) {
  const compression = raw[0];
  const length = raw.readUInt32BE(1);
  if (compression === 0) {
    return raw.subarray(5, 5 + length);
  }
  const body = raw.subarray(9, 9 + length);
  if (compression === 1) {
    return bzip2_decompress_bzh1(body);
  }
  // Python: gzip.GzipFile(fileobj=BytesIO(body)).read() -- first member only,
  // tolerant of trailing bytes. gunzipSync behaves the same way here.
  return require("zlib").gunzipSync(body);
}

function _as_container(payload) {
  // Wrap payload as an uncompressed js5 container. Re-emitting uncompressed
  // rather than recompressing keeps the bytes deterministic, which matters
  // because the advertised CRC must match exactly what goes on the wire.
  return Buffer.concat([
    Buffer.from([0]),
    (() => {
      const len = Buffer.alloc(4);
      len.writeUInt32BE(payload.length, 0);
      return len;
    })(),
    payload,
  ]);
}

function _group_table_crc_offset(table, group_id) {
  // Byte offset of group_id's CRC field inside a decompressed group table.
  let p = 0;
  const protocol = table[p];
  p += 1;
  if (protocol >= 6) {
    p += 4; // version
  }
  const flags = table[p];
  p += 1;
  const count = table.readUIntBE(p, 2);
  p += 2;
  const ids = [];
  let acc = 0;
  for (let i = 0; i < count; i++) {
    acc += table.readUIntBE(p, 2);
    p += 2;
    ids.push(acc);
  }
  if (flags & 1) {
    p += 4 * count; // name hashes
  }
  const idx = ids.indexOf(group_id);
  if (idx === -1) {
    return null;
  }
  return p + 4 * idx;
}

/** Frozen equivalent of the Js5Request dataclass. */
class Js5Request {
  constructor(priority, archive_id, group_id) {
    this.priority = priority;
    this.archive_id = archive_id;
    this.group_id = group_id;
    Object.freeze(this);
  }
}

/** sendall(): resolve when the socket has flushed the bytes to the kernel. */
function sendall(sock, buf) {
  return new Promise((resolve, reject) => {
    sock.write(buf, (err) => (err ? reject(err) : resolve()));
  });
}

/**
 * Buffered exact-length reader over a net.Socket (or any EventEmitter with
 * data/end/close/error plus write()). Resolves with exactly n bytes; rejects
 * with io.EOFError when the peer closes first -- mirroring io.py's
 * read_exact raising EOFError out of the session loop.
 */
function make_packet_reader(sock) {
  const { EOFError } = require("./io.js");
  let queue = Buffer.alloc(0);
  let dead = false;
  const waiters = [];
  const onData = (chunk) => {
    queue = queue.length === 0 ? chunk : Buffer.concat([queue, chunk]);
    pump();
  };
  const markDead = () => {
    dead = true;
    pump();
  };
  sock.on("data", onData);
  sock.on("end", markDead);
  sock.on("close", markDead);
  sock.on("error", markDead);
  function pump() {
    let i = 0;
    while (i < waiters.length) {
      const waiter = waiters[i];
      if (queue.length >= waiter.n) {
        const out = queue.subarray(0, waiter.n);
        queue = queue.subarray(waiter.n);
        waiters.splice(i, 1);
        waiter.resolve(out);
      } else if (dead) {
        waiters.splice(i, 1);
        waiter.reject(new EOFError("socket closed while reading JS5 packets"));
      } else {
        i += 1;
      }
    }
  }
  return function read_exact(n) {
    return new Promise((resolve, reject) => {
      waiters.push({ n, resolve, reject });
      pump();
    });
  };
}

/**
 * Build the unframed JS5 file-response packet for a cache entry
 * (Python Js5Session._handle_file_request lines from truncation checks to
 * packet assembly). Returns null where Python would silently return:
 * truncated entries or entries too short to hold a container header.
 */
function build_file_packet(request, raw) {
  if (request.archive_id !== 255) {
    const compression = raw[0];
    const compressed_length = raw.readUInt32BE(1);
    const container_length = (compression === 0 ? 5 : 9) + compressed_length;
    if (raw.length < container_length) {
      return null; // Python prints "truncated cache entry" and returns
    }
    var trimmed = raw;
    if (raw.length > container_length) {
      // Python strips and logs the on-disk trailer.
      trimmed = raw.subarray(0, container_length);
    }
    raw = trimmed;
  }

  if (raw.length < 5) {
    return null; // Python prints "invalid short cache entry" and returns
  }

  const compression = raw[0];
  const compressed_length = raw.readUInt32BE(1);
  let flags = compression & 0x7f;
  if (!request.priority) {
    flags |= 0x80;
  }

  const head = Buffer.alloc(10);
  head[0] = request.archive_id & 0xff;
  head.writeUInt32BE(request.group_id >>> 0, 1);
  head[5] = flags & 0xff;
  head.writeUInt32BE(compressed_length >>> 0, 6);

  return Buffer.concat([head, raw.subarray(5)]);
}

/**
 * Pure _send_framed: split a response packet into the exact sequence of
 * socket writes Python performs (first 512 bytes, then 0xFF marker plus up
 * to 511 bytes per chunk).
 */
function frame_packet(packet) {
  if (packet.length <= 512) {
    return [packet];
  }
  const chunks = [packet.subarray(0, 512)];
  let offset = 512;
  while (offset < packet.length) {
    chunks.push(Buffer.from([0xff]));
    const end = Math.min(offset + 511, packet.length);
    chunks.push(packet.subarray(offset, end));
    offset = end;
  }
  return chunks;
}

/** Apply the client-selected JS5 XOR key (opcode 4) to outgoing bytes. */
function xor_bytes(data, xor_key) {
  if (!xor_key) {
    return data;
  }
  const k = xor_key & 0xff;
  const out = Buffer.allocUnsafe(data.length);
  for (let i = 0; i < data.length; i++) out[i] = data[i] ^ k;
  return out;
}

class Js5Session {
  /**
   * sock: net.Socket (or any object with a write(callback)); cache: CacheStore;
   * peer: "ip:port" string used in log lines.
   */
  constructor(sock, cache, peer) {
    this.sock = sock;
    this.cache = cache;
    this.peer = peer;
    this.xor_key = 0;
  }

  async run_after_handshake(revision) {
    console.log(`[js5] ${this.peer} accepted revision=${revision}`);
    await sendall(this.sock, Buffer.from([0x00]));

    // tcp.js hands us the raw duplex socket as "writer"; requests are read
    // back off the same socket through our own buffered packet reader. EOF
    // surfaces as io.EOFError so tcp.js logs it like Python's tcp.py did.
    const read_packet = make_packet_reader(this.sock);

    for (;;) {
      const packet = await read_packet(6);
      const opcode = packet[0];

      if (opcode === 0 || opcode === 1) {
        const request = new Js5Request(
          opcode === 1,
          packet[1],
          packet.readUInt32BE(2)
        );
        await this._handle_file_request(request);
        continue;
      }

      if (opcode === 2) {
        console.log(`[js5] ${this.peer} priority control ${packet.toString("hex").replace(/(..)/g, "$1 ").trim()}`);
        continue;
      }

      if (opcode === 3) {
        console.log(`[js5] ${this.peer} normal control ${packet.toString("hex").replace(/(..)/g, "$1 ").trim()}`);
        continue;
      }

      if (opcode === 4) {
        this.xor_key = packet[1];
        console.log(`[js5] ${this.peer} xor-key=${this.xor_key}`);
        continue;
      }

      if (opcode === 6) {
        console.log(`[js5] ${this.peer} setup ${packet.toString("hex").replace(/(..)/g, "$1 ").trim()}`);
        continue;
      }

      if (opcode === 7) {
        console.log(`[js5] ${this.peer} disconnect control`);
        return;
      }

      console.log(
        `[js5] ${this.peer} unknown opcode=${opcode} ${packet.toString("hex").replace(/(..)/g, "$1 ").trim()}`
      );
    }
  }

  async _patched_read(archive_id, group_id) {
    // cache.read() with substituted groups and reconciled CRCs applied.
    //
    // Every read path must use this. Reading the cache directly anywhere
    // re-introduces the mismatch this exists to remove.
    const raw = this.cache.read(archive_id, group_id);

    // Level 0: the substituted group itself.
    if (raw === null) {
      const substitute = _load_substitute(archive_id, group_id);
      if (substitute !== null) {
        console.log(
          `[js5] ${this.peer} SUBSTITUTE archive=${archive_id} ` +
            `group=${group_id} bytes=${substitute.length} ` +
            `crc=0x${crc32(substitute).toString(16).padStart(8, "0")}`
        );
        return substitute;
      }
      return null;
    }

    // Levels 1 and 2 are DISABLED. Rewriting the recorded CRC does not work:
    // the master index (255/255) is *signed*, and it covers the group-table
    // CRCs. Patching a group table forces a matching edit to the master
    // index, which invalidates the signature -- the client verifies it,
    // throws RuntimeException and dies with error_game_crash before it ever
    // reaches the login screen.
    //
    // The workable direction is the reverse: forge the substitute so its CRC
    // equals the value already recorded (0xaacfba29 for archive 6 group 1).
    // CRC32 is affine over GF(2), so 32 probe CRCs are enough to solve for
    // four bytes that land on any target. Put them somewhere inert -- the
    // palette, not pixel indices, which could go out of bounds and crash the
    // sprite reader. Then nothing signed has to change.
    return raw;

    if (archive_id !== 255) {
      return raw;
    }

    // Level 1: a group table vouching for a substituted group.
    if (group_id !== 255) {
      let patched = raw;
      for (const [key, _name] of _SUBSTITUTES) {
        const [arch, grp] = key.split(",").map(Number);
        if (arch !== group_id) {
          continue;
        }
        const substitute = _load_substitute(arch, grp);
        if (substitute === null) {
          continue;
        }
        const table = Buffer.from(_decompress(patched));
        const off = _group_table_crc_offset(table, grp);
        if (off === null) {
          console.log(`[js5] ${this.peer} group ${grp} absent from table ${arch}; CRC not patched`);
          continue;
        }
        const want = crc32(substitute) >>> 0;
        const had = table.readUInt32BE(off);
        table.writeUInt32BE(want, off);
        patched = _as_container(table);
        console.log(
          `[js5] ${this.peer} patched table ${arch} group ${grp} ` +
            `crc 0x${had.toString(16).padStart(8, "0")} -> 0x${want.toString(16).padStart(8, "0")}`
        );
      }
      return patched;
    }

    // Level 2: the master index vouching for those group tables.
    const touched = new Set();
    for (const key of _SUBSTITUTES.keys()) touched.add(Number(key.split(",")[0]));
    if (touched.size === 0) {
      return raw;
    }
    const body = Buffer.from(_decompress(raw));
    const count = body[0];
    for (const arch of [...touched].sort((a, b) => a - b)) {
      if (arch >= count) {
        continue;
      }
      const table = await this._patched_read(255, arch);
      if (table === null) {
        continue;
      }
      const off = 1 + arch * 72; // crc(4) + version(4) + whirlpool(64)
      const want = crc32(table) >>> 0;
      const had = body.readUInt32BE(off);
      if (had === want) {
        continue;
      }
      body.writeUInt32BE(want, off);
      console.log(
        `[js5] ${this.peer} patched master index archive ${arch} ` +
          `crc 0x${had.toString(16).padStart(8, "0")} -> 0x${want.toString(16).padStart(8, "0")}`
      );
    }
    return _as_container(body);
  }

  async _build_master_index() {
    // Generate the js5 master index (archive 255, group 255) from the
    // per-archive indexes (255, N). The master index is never stored on disk;
    // a real js5 server computes it live. Format:
    //   container([count:u8] then per archive [crc:u32][version:u32][whirlpool:64b])
    const idx_path = path.join(String(this.cache.cache_dir), "main_file_cache.idx255");
    let count = 0;
    try {
      if (fs.statSync(idx_path).isFile()) {
        count = Math.floor(fs.statSync(idx_path).size / 6);
      }
    } catch {
      count = 0;
    }
    const body = Buffer.alloc(1 + count * 72);
    body[0] = count & 0xff;
    for (let n = 0; n < count; n++) {
      const entry = await this._patched_read(255, n);
      const crcValue = entry === null ? 0 : crc32(entry) >>> 0;
      body.writeUInt32BE(crcValue, 1 + n * 72); // crc
      body.writeUInt32BE(0, 1 + n * 72 + 4); // version = 0
      // whirlpool stays zero-filled (64 bytes)
    }
    return _as_container(body); // compression byte 0 + u32 length + body
  }

  async _handle_file_request(request) {
    let raw;
    if (request.archive_id === 255 && request.group_id === 255) {
      raw = await this._patched_read(255, 255);
      if (raw === null) {
        raw = await this._build_master_index();
        console.log(`[js5] ${this.peer} generated unsigned fallback master index (${raw.length}b)`);
      } else {
        console.log(`[js5] ${this.peer} loaded signed master index (${raw.length}b)`);
      }
    } else {
      raw = await this._patched_read(request.archive_id, request.group_id);
    }
    if (raw === null) {
      console.log(
        `[js5] ${this.peer} MISS archive=${request.archive_id} ` +
          `group=${request.group_id} priority=${request.priority}`
      );
      // Answer the miss instead of staying silent. A bare return leaves the
      // client waiting ~30s forever; see the long comment in js5.py. Reply
      // with a well-formed empty container (compression 0, length 0).
      let flags = 0;
      if (!request.priority) {
        flags |= 0x80;
      }
      const empty = Buffer.alloc(10);
      empty[0] = request.archive_id & 0xff;
      empty.writeUInt32BE(request.group_id >>> 0, 1);
      empty[5] = flags & 0xff;
      empty.writeUInt32BE(0, 6); // compressed length 0
      await this._send_framed(empty);
      console.log(
        `[js5] ${this.peer} sent EMPTY container for miss ` +
          `archive=${request.archive_id} group=${request.group_id}`
      );
      return;
    }

    const packet = build_file_packet(request, raw);
    if (packet === null) {
      // Python's branch order means archive != 255 nulls are always the
      // "truncated" case (its slice-based header read tolerates short
      // entries); only archive 255 entries can hit "invalid short".
      if (request.archive_id !== 255) {
        console.log(
          `[js5] ${this.peer} truncated cache entry archive=${request.archive_id} ` +
            `group=${request.group_id} bytes=${raw.length} expected=${expected_container_length(raw)}`
        );
      } else {
        console.log(`[js5] ${this.peer} invalid short cache entry: ${JSON.stringify(request)}`);
      }
      return;
    }

    // Trailer logging parity: report what got stripped from over-long entries.
    if (request.archive_id !== 255) {
      const container_length = expected_container_length(raw);
      if (container_length !== null && raw.length > container_length) {
        const trailer = raw.subarray(container_length);
        console.log(
          `[js5] ${this.peer} stripped disk trailer=${trailer.toString("hex").replace(/(..)/g, "$1 ").trim()} ` +
            `archive=${request.archive_id} group=${request.group_id}`
        );
      }
    }

    await this._send_framed(packet);
    console.log(
      `[js5] ${this.peer} sent archive=${request.archive_id} ` +
        `group=${request.group_id} bytes=${raw.length} priority=${request.priority}`
    );
  }

  async _send_framed(packet) {
    for (const chunk of frame_packet(packet)) {
      await sendall(this.sock, xor_bytes(chunk, this.xor_key));
    }
  }
}

/** Container length implied by a raw entry header. Mirrors Python's
 * slice-based int.from_bytes(raw[1:5]), which tolerates short entries instead
 * of throwing -- needed so truncated/short classification matches exactly. */
function expected_container_length(raw) {
  let compressed_length = 0;
  const stop = Math.min(5, raw.length);
  for (let i = 1; i < stop; i++) compressed_length = compressed_length * 256 + raw[i];
  return (raw[0] === 0 ? 5 : 9) + compressed_length;
}

module.exports = {
  Js5Request,
  Js5Session,
  _load_substitute,
  _decompress,
  _as_container,
  _group_table_crc_offset,
  build_file_packet,
  frame_packet,
  xor_bytes,
  crc32,
  SUBSTITUTE_DIR,
  _SUBSTITUTES,
};
